hbase/hbase-bindings/src/main/java/cz/o2/proxima/storage/hbase/HBaseLogObservable.java

```diff
@@ -0,0 +1,167 @@
+/**
+ * Copyright 2017 O2 Czech Republic, a.s.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.o2.proxima.storage.hbase;
+
+import cz.o2.proxima.repository.AttributeDescriptor;
+import cz.o2.proxima.repository.EntityDescriptor;
+import cz.o2.proxima.storage.Partition;
+import cz.o2.proxima.storage.StreamElement;
+import cz.o2.proxima.storage.batch.BatchLogObservable;
+import cz.o2.proxima.storage.batch.BatchLogObserver;
+import java.io.IOException;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.CellScanner;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.ResultScanner;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.hadoop.hbase.filter.BinaryComparator;
+import org.apache.hadoop.hbase.filter.ColumnPrefixFilter;
+import org.apache.hadoop.hbase.filter.CompareFilter;
+import org.apache.hadoop.hbase.filter.Filter;
+import org.apache.hadoop.hbase.filter.FilterList;
+import org.apache.hadoop.hbase.filter.QualifierFilter;
+
+/**
+ * A {@code BatchLogObservable} for HBase.
+ */
+class HBaseLogObservable extends HBaseClientWrapper implements BatchLogObservable {
+
+  private final EntityDescriptor entity;
+
+  public HBaseLogObservable(
+      URI uri, Configuration conf, Map<String, Object> cfg,
+      EntityDescriptor entity) {
+
+    super(uri, conf, cfg);
+    this.entity = entity;
+  }
+
+  @Override
+  public List<Partition> getPartitions(long startStamp, long endStamp) {
+    try {
+      List<Partition> ret = new ArrayList<>();
+      byte[][] end = conn.getRegionLocator(table).getEndKeys();
+      byte[] startPos = new byte[0];
+      for (int i = 0; i < end.length; i++) {
+        ret.add(new HBasePartition(i, startPos, end[i], startStamp, endStamp));
+        startPos = end[i];
+      }
+      ret.add(new HBasePartition(
+          end.length, startPos, new byte[0], startStamp, endStamp));
+      return ret;
+    } catch (IOException ex) {
+      throw new RuntimeException(ex);
+    }
+  }
+
+  @Override
+  public void observe(
+      List<Partition> partitions,
+      List<AttributeDescriptor<?>> attributes,
+      BatchLogObserver observer) {
+
+    Thread consumer = new Thread(() -> {
```

Comment:
I guess this should be done in more generic way. Isn't this similiar for other storages?

Reply:
how about having a bounded executor service (shared accross batch observables) for this? This could potentially start unlimited num of consumer at the same time

Reply:
Absolutely agree. Filled #35.

Reply:
👍